docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+---
+title: "Blocks storage sharding"
+linkTitle: "Blocks storage sharding"
+weight: 1
+slug: blocks-storage-sharding
+---
+
+- Author: [Marco Pracucci](https://github.com/pracucci)
+- Date: March 2020
+- Status: proposal
+
+## Problem
+
+In Cortex, when using the experimental blocks storage, each querier internally runs the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). This means that each querier has a full view over all blocks in the long-term storage and all blocks index headers are loaded in each querier memory. The querier memory usage linearly increase with number and size of all blocks in the storage, imposing a scalability limit to the blocks storage.
+
+In this proposal we want to solve this. In particular, we want to:
+
+1. Shard blocks (index headers) across a pool of nodes
+2. Do not compromise HA on the read path (if a node fails, queries should continue to work)
+3. Do not compromise correctness (either the query result is correct or it fails)
+
+## Proposed solution
+
+The idea is to introduce a new Cortex service - `store-gateway` - internally running the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). At query time, a querier will run a query fetching the matching series both from ingesters and the subset of gateways holding the related blocks (based on the query time range). Blocks are replicated across the gateways in order to guarantee query results consistency and HA even in the case of a gateway instance failure.
+
+### Ring-based sharding and replication
+
+In order to build blocks sharding and replication, the `store-gateway` instances form a [ring](../architecture.md#the-hash-ring). Each gateway instance uses a custom [`MetaFetcherFilter`](https://github.com/thanos-io/thanos/blob/master/pkg/block/fetcher.go#L108) to filter blocks loaded on the instance itself, keeping only blocks whose `hash(block-id)` is within the tokens range assigned to the gateway instance within the ring.
```

Comment:
If each `store-gateway` is holding the blocks assigned to it, would that mean that we'd be storing all of S3, in these gateways? Or does the `MetaFectherFilter` only store the metadata for it's assigned blocks, and then caches recently read blocks? If so that could be made more clear here.

Reply:
The `MetaFetcherFilter` filters out blocks not assigned to a specific `store-gateway` shard. For the matching blocks, only the binary block index header is downloaded. The full index and chunks never hit the `store-gateway` disk and they're directly streamed from the storage (S3/GCS) using byte-range requests (which is how it works right now).